File: js/tilings/trapezoid-triangle-weave.js
/**
 * Generates a trapezoid-triangle weave board as a graph.
 *
 * The tiling is built on a unit triangular lattice.  Each triangular unit has
 * a side-3 equilateral triangle surrounded by three isosceles trapezoids.
 * The trapezoid edge touching the triangle is length 4, with one endpoint
 * aligned to the triangle side and one extra unit segment extending past the
 * other endpoint.  The opposite trapezoid edge is length 5, and the two short
 * legs are length 1.  Mirrored up/down units share those length-5 outer edges
 * to create the woven pattern.  Longer edges are subdivided into unit segments
 * so graph neighbors can be matched across partial edges.
 */
export function generateTrapezoidTriangleWeaveBoard(options) {
  const { cols, rows, tileSize: side, colorCount, rng } = options;
  const periodI = 6;
  const periodJ = 6;
  const targetWidth = Math.max(cols, 1) * 3.0 * side;
  const targetHeight = Math.max(rows, 1) * Math.sqrt(3) * side;
  const pad = 4;
  const cropMarginX = periodI * side / 2;
  const cropMarginY = periodJ * Math.sqrt(3) * side / 4;

  const upUnit = [
    { type: 'triangle-up', corners: [[0, 0], [3, 0], [0, 3]] },
    { type: 'trapezoid', corners: [[0, 0], [4, 0], [5, -1], [0, -1]] },
    { type: 'trapezoid', corners: [[3, 0], [-1, 4], [-1, 5], [4, 0]] },
    { type: 'trapezoid', corners: [[0, 3], [0, -1], [-1, -1], [-1, 4]] }
  ];

  const downUnit = upUnit.map(tile => ({
    type: tile.type === 'triangle-up' ? 'triangle-down' : tile.type,
    corners: tile.corners.map(([i, j]) => reflectAcrossHorizontalLatticeLine(i, j, -1))
  }));
  const unitTemplates = [upUnit, downUnit];

  const unitGroups = [];
  const repeatMinY = -pad;
  const repeatMaxY = Math.ceil(targetHeight / (periodJ * Math.sqrt(3) * side / 2)) + pad;
  const shearPad = Math.ceil(Math.max(0, repeatMaxY * periodJ) / (2 * periodI)) + 2;
  const reverseShearPad = Math.ceil(Math.max(0, -repeatMinY * periodJ) / (2 * periodI)) + 2;
  const repeatMinX = -pad - shearPad;
  const repeatMaxX = Math.ceil(targetWidth / (periodI * side)) + pad + reverseShearPad;

  for (let ry = repeatMinY; ry <= repeatMaxY; ry++) {
    for (let rx = repeatMinX; rx <= repeatMaxX; rx++) {
      const offsetI = rx * periodI;
      const offsetJ = ry * periodJ;
      for (const unit of unitTemplates) {
        const orientation = unit[0].type === 'triangle-up' ? 'up' : 'down';
        const shiftedCorners = unit.map(tile => tile.corners.map(([i, j]) => [i + offsetI, j + offsetJ]));
        const unitTiles = unit.map((tile, index) => makeRawTile(tile.type, subdividedPolygonKeys(shiftedCorners[index])));
        const triangle = unitTiles.find(tile => tile.type.startsWith('triangle'));
        const anchor = averagePoint(triangle.points);
        if (anchor[0] >= -cropMarginX && anchor[0] <= targetWidth + cropMarginX &&
            anchor[1] >= -cropMarginY && anchor[1] <= targetHeight + cropMarginY) {
          const allCorners = shiftedCorners.flat();
          unitGroups.push({
            orientation,
            minJ: Math.min(...allCorners.map(([, j]) => j)),
            maxJ: Math.max(...allCorners.map(([, j]) => j)),
            tiles: unitTiles
          });
        }
      }
    }
  }

  // Keep whole triangle+three-trapezoid units, but choose horizontal lattice
  // lines where the top row of up units and bottom row of down units have
  // flat outer trapezoid edges.  This avoids the single-vertex "spikes" that
  // appear when the crop starts on a down-unit apex or ends on an up-unit apex.
  const topFlatJ = Math.min(...unitGroups.filter(group => group.orientation === 'up').map(group => group.minJ));
  const bottomFlatJ = Math.max(...unitGroups.filter(group => group.orientation === 'down').map(group => group.maxJ));
  const rawTiles = unitGroups
    .filter(group => group.minJ >= topFlatJ && group.maxJ <= bottomFlatJ)
    .flatMap(group => group.tiles);

  const cropped = rawTiles
    .map(tile => ({ ...tile, centroid: averagePoint(tile.points) }))
    .sort((a, b) => {
      if (Math.abs(a.centroid[1] - b.centroid[1]) > 1e-6) return a.centroid[1] - b.centroid[1];
      return a.centroid[0] - b.centroid[0];
    });

  const tiles = cropped.map((tile, id) => ({
    id,
    type: tile.type,
    colorId: Math.floor(rng() * colorCount),
    ownerId: null,
    points: tile.points.map(point => [...point]),
    neighbors: []
  }));

  const neighborSets = tiles.map(() => new Set());
  const edgeToTiles = new Map();
  cropped.forEach((tile, tileId) => {
    for (let i = 0; i < tile.vertexKeys.length; i++) {
      const a = tile.vertexKeys[i];
      const b = tile.vertexKeys[(i + 1) % tile.vertexKeys.length];
      const key = canonicalPairKey(a, b);
      if (!edgeToTiles.has(key)) edgeToTiles.set(key, []);
      edgeToTiles.get(key).push(tileId);
    }
  });

  for (const tileIds of edgeToTiles.values()) {
    if (tileIds.length < 2) continue;
    for (let i = 0; i < tileIds.length; i++) {
      for (let j = i + 1; j < tileIds.length; j++) {
        const a = tileIds[i];
        const b = tileIds[j];
        if (a !== b) {
          neighborSets[a].add(b);
          neighborSets[b].add(a);
        }
      }
    }
  }

  tiles.forEach((tile, id) => {
    tile.neighbors = [...neighborSets[id]].sort((a, b) => a - b);
  });

  let bbox = computeBoundingBox(tiles);
  for (const tile of tiles) {
    tile.points = tile.points.map(([x, y]) => [x - bbox.minX, y - bbox.minY]);
  }
  bbox = computeBoundingBox(tiles);

  const startTileIds = uniqueIds([
    closestTileId(tiles, 0, 0),
    closestTileId(tiles, bbox.maxX, bbox.maxY),
    closestTileId(tiles, bbox.maxX, 0),
    closestTileId(tiles, 0, bbox.maxY)
  ]);

  return {
    version: 1,
    generator: 'trapezoid-triangle-weave',
    width: bbox.maxX,
    height: bbox.maxY,
    cols,
    rows,
    tiles,
    startTileIds
  };

  function makeRawTile(type, vertexKeys) {
    return {
      type,
      vertexKeys,
      points: vertexKeys.map(key => latticePointFromKey(key, side))
    };
  }
}

function reflectAcrossHorizontalLatticeLine(i, j, lineJ) {
  return [i + j - lineJ, 2 * lineJ - j];
}

function subdividedPolygonKeys(corners) {
  const keys = [];
  for (let sideIdx = 0; sideIdx < corners.length; sideIdx++) {
    const [ai, aj] = corners[sideIdx];
    const [bi, bj] = corners[(sideIdx + 1) % corners.length];
    const steps = Math.max(Math.abs(bi - ai), Math.abs(bj - aj));
    const stepI = (bi - ai) / steps;
    const stepJ = (bj - aj) / steps;
    for (let step = 0; step < steps; step++) {
      keys.push(latticeKey(ai + step * stepI, aj + step * stepJ));
    }
  }
  return keys;
}

function latticeKey(i, j) {
  return `${i},${j}`;
}

function latticePointFromKey(key, side) {
  const [i, j] = key.split(',').map(Number);
  return [side * (i + j / 2), side * Math.sqrt(3) * j / 2];
}

function canonicalPairKey(a, b) {
  return a < b ? `${a}~${b}` : `${b}~${a}`;
}

function averagePoint(points) {
  const sum = points.reduce((acc, [x, y]) => [acc[0] + x, acc[1] + y], [0, 0]);
  return [sum[0] / points.length, sum[1] / points.length];
}

function computeBoundingBox(tiles) {
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;

  for (const tile of tiles) {
    for (const [x, y] of tile.points) {
      minX = Math.min(minX, x);
      minY = Math.min(minY, y);
      maxX = Math.max(maxX, x);
      maxY = Math.max(maxY, y);
    }
  }

  return { minX, minY, maxX, maxY };
}

function closestTileId(tiles, tx, ty) {
  let bestId = tiles[0]?.id ?? 0;
  let bestDist = Infinity;
  for (const tile of tiles) {
    const [cx, cy] = averagePoint(tile.points);
    const dist = (cx - tx) ** 2 + (cy - ty) ** 2;
    if (dist < bestDist) {
      bestDist = dist;
      bestId = tile.id;
    }
  }
  return bestId;
}

function uniqueIds(ids) {
  return [...new Set(ids.filter(id => Number.isInteger(id)))];
}

File: go-flood/pkg/tilings/trapezoid_triangle_weave.go
package tilings

import (
	"fmt"
	"go-flood/pkg/core"
	"math"
	"sort"
	"strings"
)

type trapezoidTriangleWeaveRawTile struct {
	TileType   string
	VertexKeys []string
	Points     []core.Point
	Centroid   core.Point
}

type trapezoidTriangleWeaveMotifTile struct {
	TileType string
	Corners  [][2]int
}

type trapezoidTriangleWeaveUnitGroup struct {
	Orientation string
	MinJ        int
	MaxJ        int
	Tiles       []trapezoidTriangleWeaveRawTile
}

// GenerateTrapezoidTriangleWeaveBoard generates a triangular-lattice weave
// made from side-3 equilateral triangles and isosceles trapezoids.  Each
// triangle is surrounded by three trapezoids whose inner edge has length 4,
// opposite edge has length 5, and two short legs have length 1.  The inner
// length-4 trapezoid edge is offset against the triangle side so one endpoint
// aligns with the triangle and one unit segment extends past the other
// endpoint.  Mirrored up/down units share the length-5 outer edges to create
// the woven pattern. Longer edges are subdivided into unit segments so graph
// neighbors can be matched across partial edges.
func GenerateTrapezoidTriangleWeaveBoard(options Options) core.Board {
	side := options.TileSize
	const periodI = 6
	const periodJ = 6
	targetWidth := math.Max(float64(options.Cols), 1) * 3.0 * side
	targetHeight := math.Max(float64(options.Rows), 1) * math.Sqrt(3) * side
	pad := 4
	cropMarginX := periodI * side / 2
	cropMarginY := periodJ * math.Sqrt(3) * side / 4

	upUnit := []trapezoidTriangleWeaveMotifTile{
		{TileType: "triangle-up", Corners: [][2]int{{0, 0}, {3, 0}, {0, 3}}},
		{TileType: "trapezoid", Corners: [][2]int{{0, 0}, {4, 0}, {5, -1}, {0, -1}}},
		{TileType: "trapezoid", Corners: [][2]int{{3, 0}, {-1, 4}, {-1, 5}, {4, 0}}},
		{TileType: "trapezoid", Corners: [][2]int{{0, 3}, {0, -1}, {-1, -1}, {-1, 4}}},
	}

	downUnit := []trapezoidTriangleWeaveMotifTile{}
	for _, tile := range upUnit {
		reflected := trapezoidTriangleWeaveMotifTile{TileType: tile.TileType}
		if tile.TileType == "triangle-up" {
			reflected.TileType = "triangle-down"
		}
		reflected.Corners = make([][2]int, len(tile.Corners))
		for i, corner := range tile.Corners {
			reflected.Corners[i] = trapezoidTriangleWeaveReflectAcrossHorizontalLatticeLine(corner[0], corner[1], -1)
		}
		downUnit = append(downUnit, reflected)
	}
	unitTemplates := [][]trapezoidTriangleWeaveMotifTile{upUnit, downUnit}

	makeRawTile := func(tileType string, vertexKeys []string) trapezoidTriangleWeaveRawTile {
		points := make([]core.Point, len(vertexKeys))
		for i, key := range vertexKeys {
			points[i] = trapezoidTriangleWeaveLatticePointFromKey(key, side)
		}
		return trapezoidTriangleWeaveRawTile{
			TileType:   tileType,
			VertexKeys: vertexKeys,
			Points:     points,
		}
	}

	unitGroups := []trapezoidTriangleWeaveUnitGroup{}
	repeatMinY := -pad
	repeatMaxY := int(math.Ceil(targetHeight/(periodJ*math.Sqrt(3)*side/2))) + pad
	shearPad := int(math.Ceil(math.Max(0, float64(repeatMaxY*periodJ))/(2*periodI))) + 2
	reverseShearPad := int(math.Ceil(math.Max(0, float64(-repeatMinY*periodJ))/(2*periodI))) + 2
	repeatMinX := -pad - shearPad
	repeatMaxX := int(math.Ceil(targetWidth/(periodI*side))) + pad + reverseShearPad

	for ry := repeatMinY; ry <= repeatMaxY; ry++ {
		for rx := repeatMinX; rx <= repeatMaxX; rx++ {
			offsetI := rx * periodI
			offsetJ := ry * periodJ
			for _, unit := range unitTemplates {
				orientation := "down"
				if unit[0].TileType == "triangle-up" {
					orientation = "up"
				}

				unitTiles := []trapezoidTriangleWeaveRawTile{}
				var triangle *trapezoidTriangleWeaveRawTile
				minJ := maxIntValue()
				maxJ := -maxIntValue()
				for _, tile := range unit {
					corners := make([][2]int, len(tile.Corners))
					for i, corner := range tile.Corners {
						corners[i] = [2]int{corner[0] + offsetI, corner[1] + offsetJ}
						minJ = minInt(minJ, corners[i][1])
						maxJ = maxInt(maxJ, corners[i][1])
					}
					raw := makeRawTile(tile.TileType, trapezoidTriangleWeaveSubdividedPolygonKeys(corners))
					unitTiles = append(unitTiles, raw)
					if strings.HasPrefix(raw.TileType, "triangle") {
						triangle = &unitTiles[len(unitTiles)-1]
					}
				}
				if triangle == nil {
					continue
				}
				anchor := trapezoidTriangleWeaveAveragePoint(triangle.Points)
				if anchor[0] >= -cropMarginX && anchor[0] <= targetWidth+cropMarginX && anchor[1] >= -cropMarginY && anchor[1] <= targetHeight+cropMarginY {
					unitGroups = append(unitGroups, trapezoidTriangleWeaveUnitGroup{Orientation: orientation, MinJ: minJ, MaxJ: maxJ, Tiles: unitTiles})
				}
			}
		}
	}

	// Keep whole triangle+three-trapezoid units, but choose horizontal lattice
	// lines where the top row of up units and bottom row of down units have
	// flat outer trapezoid edges. This avoids the single-vertex spikes that
	// appear when the crop starts on a down-unit apex or ends on an up-unit apex.
	topFlatJ := maxIntValue()
	bottomFlatJ := -maxIntValue()
	for _, group := range unitGroups {
		if group.Orientation == "up" {
			topFlatJ = minInt(topFlatJ, group.MinJ)
		}
		if group.Orientation == "down" {
			bottomFlatJ = maxInt(bottomFlatJ, group.MaxJ)
		}
	}

	rawTiles := []trapezoidTriangleWeaveRawTile{}
	for _, group := range unitGroups {
		if group.MinJ >= topFlatJ && group.MaxJ <= bottomFlatJ {
			rawTiles = append(rawTiles, group.Tiles...)
		}
	}

	cropped := []trapezoidTriangleWeaveRawTile{}
	for _, tile := range rawTiles {
		tile.Centroid = trapezoidTriangleWeaveAveragePoint(tile.Points)
		cropped = append(cropped, tile)
	}

	sort.Slice(cropped, func(i, j int) bool {
		if math.Abs(cropped[i].Centroid[1]-cropped[j].Centroid[1]) > 1e-6 {
			return cropped[i].Centroid[1] < cropped[j].Centroid[1]
		}
		return cropped[i].Centroid[0] < cropped[j].Centroid[0]
	})

	tiles := make([]core.Tile, len(cropped))
	for id, tile := range cropped {
		points := make([]core.Point, len(tile.Points))
		copy(points, tile.Points)
		tiles[id] = core.Tile{
			ID:        id,
			ColorID:   int(options.RNG() * float64(options.ColorCount)),
			OwnerID:   nil,
			Points:    points,
			Neighbors: []int{},
		}
	}

	neighborSets := make([]map[int]bool, len(tiles))
	for i := range neighborSets {
		neighborSets[i] = map[int]bool{}
	}

	edgeToTiles := map[string][]int{}
	for tileID, tile := range cropped {
		for i, a := range tile.VertexKeys {
			b := tile.VertexKeys[(i+1)%len(tile.VertexKeys)]
			key := trapezoidTriangleWeaveCanonicalPairKey(a, b)
			edgeToTiles[key] = append(edgeToTiles[key], tileID)
		}
	}

	for _, tileIDs := range edgeToTiles {
		if len(tileIDs) < 2 {
			continue
		}
		for i := 0; i < len(tileIDs); i++ {
			for j := i + 1; j < len(tileIDs); j++ {
				a := tileIDs[i]
				b := tileIDs[j]
				if a != b {
					neighborSets[a][b] = true
					neighborSets[b][a] = true
				}
			}
		}
	}

	for id := range tiles {
		neighbors := make([]int, 0, len(neighborSets[id]))
		for neighborID := range neighborSets[id] {
			neighbors = append(neighbors, neighborID)
		}
		sort.Ints(neighbors)
		tiles[id].Neighbors = neighbors
	}

	minX, minY, _, _ := trapezoidTriangleWeaveBoundingBox(tiles)
	for i := range tiles {
		for j, p := range tiles[i].Points {
			tiles[i].Points[j] = core.Point{p[0] - minX, p[1] - minY}
		}
	}

	_, _, width, height := trapezoidTriangleWeaveBoundingBox(tiles)
	startTileIds := trapezoidTriangleWeaveUniqueIds([]int{
		trapezoidTriangleWeaveClosestTileID(tiles, 0, 0),
		trapezoidTriangleWeaveClosestTileID(tiles, width, height),
		trapezoidTriangleWeaveClosestTileID(tiles, width, 0),
		trapezoidTriangleWeaveClosestTileID(tiles, 0, height),
	})

	return core.Board{
		Version:      1,
		Generator:    "trapezoid-triangle-weave",
		Width:        width,
		Height:       height,
		Cols:         options.Cols,
		Rows:         options.Rows,
		Tiles:        tiles,
		StartTileIds: startTileIds,
	}
}

func trapezoidTriangleWeaveReflectAcrossHorizontalLatticeLine(i, j, lineJ int) [2]int {
	return [2]int{i + j - lineJ, 2*lineJ - j}
}

func trapezoidTriangleWeaveSubdividedPolygonKeys(corners [][2]int) []string {
	keys := []string{}
	for sideIdx := 0; sideIdx < len(corners); sideIdx++ {
		a := corners[sideIdx]
		b := corners[(sideIdx+1)%len(corners)]
		steps := int(math.Max(math.Abs(float64(b[0]-a[0])), math.Abs(float64(b[1]-a[1]))))
		stepI := (b[0] - a[0]) / steps
		stepJ := (b[1] - a[1]) / steps
		for step := 0; step < steps; step++ {
			keys = append(keys, trapezoidTriangleWeaveLatticeKey(a[0]+step*stepI, a[1]+step*stepJ))
		}
	}
	return keys
}

func trapezoidTriangleWeaveLatticeKey(i, j int) string {
	return fmt.Sprintf("%d,%d", i, j)
}

func trapezoidTriangleWeaveLatticePointFromKey(key string, side float64) core.Point {
	parts := strings.Split(key, ",")
	if len(parts) != 2 {
		return core.Point{0, 0}
	}
	var i, j int
	_, _ = fmt.Sscanf(parts[0], "%d", &i)
	_, _ = fmt.Sscanf(parts[1], "%d", &j)
	return core.Point{
		side * (float64(i) + float64(j)/2),
		side * math.Sqrt(3) * float64(j) / 2,
	}
}

func trapezoidTriangleWeaveCanonicalPairKey(a, b string) string {
	if a < b {
		return a + "~" + b
	}
	return b + "~" + a
}

func trapezoidTriangleWeaveAveragePoint(points []core.Point) core.Point {
	sumX := 0.0
	sumY := 0.0
	for _, point := range points {
		sumX += point[0]
		sumY += point[1]
	}
	return core.Point{sumX / float64(len(points)), sumY / float64(len(points))}
}

func trapezoidTriangleWeaveBoundingBox(tiles []core.Tile) (float64, float64, float64, float64) {
	minX := math.MaxFloat64
	minY := math.MaxFloat64
	maxX := -math.MaxFloat64
	maxY := -math.MaxFloat64
	for _, tile := range tiles {
		for _, p := range tile.Points {
			minX = math.Min(minX, p[0])
			minY = math.Min(minY, p[1])
			maxX = math.Max(maxX, p[0])
			maxY = math.Max(maxY, p[1])
		}
	}
	return minX, minY, maxX, maxY
}

func trapezoidTriangleWeaveClosestTileID(tiles []core.Tile, tx, ty float64) int {
	bestID := -1
	bestDist := math.MaxFloat64
	for _, tile := range tiles {
		center := trapezoidTriangleWeaveAveragePoint(tile.Points)
		dist := math.Hypot(center[0]-tx, center[1]-ty)
		if dist < bestDist {
			bestDist = dist
			bestID = tile.ID
		}
	}
	return bestID
}

func trapezoidTriangleWeaveUniqueIds(ids []int) []int {
	seen := map[int]bool{}
	unique := []int{}
	for _, id := range ids {
		if id < 0 || seen[id] {
			continue
		}
		seen[id] = true
		unique = append(unique, id)
	}
	return unique
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func maxIntValue() int {
	return int(^uint(0) >> 1)
}
